### DISCUSSION TITLE
Circular dependency problem with self-mutating CDK pipeline and ECS Fargate

### DISCUSSION THREAD

**Discussion by suavelizard** [aws/aws-cdk#22947] [Q&A]
*TL;DR All I want to is to write CDK code to create a service + pipeline to deploy the service but cannot because of circular dependency*

I have a CDK project with 2 stacks

- Stack A: PipelineStack
- Stack B: ApplicationStack

This CDK project is attempting to do the following
1 - Create all resources via CDK (ECS cluster app, Loadbalancer, ECR, Pipeline etc.)
2 - Deploy changes in a self mutating pipeline

The problems arise when I attempt to do the following:
` new PipelineStack()`
    -> Pulls CDK code
    -> Creates ECR Repo
    -> Creates Pipeline
    -> Adds ECRDeployment action

`new ApplicationStack()`
    -> Creates ECS Service
    -> Reads from ECR
    -> Runs Container with ECR image

I encounter an issue where the deployment and the creation of the services conflict:
```
 new codepipeline_actions.EcsDeployAction({
              actionName: "ECSDeploy_Action",
              input: appSourceOutput,
              service: <How to reference service before it is created>,

```
and 
```
FargateTask.addContainer(`${namePrefix}-Container`, {

      image: <FROM ECR Repository That doesn't exist yet>
    });
```

The service needs the ECR respository to exist and the ECSDeploy needs the service to exist.
How can I automate this?  I havent found any examples which is surprising since deploying a self mutating pipeline that updates an ECS service seems like a common usecase.

Solutions like [this](https://github.com/flochaz/aws-cdk/blob/master/packages/%40aws-cdk/aws-codepipeline-actions/test/integ.pipeline-ecs-separate-source.lit.ts) do not seem to work in CDKV2 (PR here: https://github.com/flochaz/aws-cdk/commit/3fdb78e828222df529e4ccd81c90120a5107f650#diff-774d4642172b82145ae731a64631ebcbbde6d0af076c8ea23bb8cb33c21aa400)

Similar: 
* https://github.com/aws/aws-cdk/issues/7746#issuecomment-622945134
* https://github.com/KierenMartin/payslip-cdk/blob/edb865c3d1a4a27c575226b3a6d980d872e9e477/lib/cdk-automation-stack.ts#L187
* https://github.com/awslabs/ecs-refarch-continuous-deployment/issues/33
* https://github.com/aws/containers-roadmap/issues/631

**Comment by suavelizard** [comment]
Can Lazy be used to get around this?
